### PR TITLE
Update startup

### LINKED ISSUE #181: The startup scripts need to output data in the new dict format
At the moment (after the merger of #180) I get this exception:

```
make[1]: Leaving directory '/home/ltratt/warmup_experiment/benchmarks'
python2.7 krun/krun.py startup.krun
[2016-10-05 17:36:24: INFO] Wait 60 secs to allow system to cool prior to collecting initial temperature readings
Traceback (most recent call last):
  File "krun/krun.py", line 346, in <module>
    main(parser)
  File "krun/krun.py", line 209, in main
    inner_main(mailer, config, args)
  File "krun/krun.py", line 295, in inner_main
    sanity_checks(config, platform)
  File "krun/krun.py", line 68, in sanity_checks
    vm_info["vm_def"].sanity_checks()
  File "/home/ltratt/warmup_experiment/krun/krun/vm_defs.py", line 633, in sanity_checks
    self._check_truffle_enabled()
  File "/home/ltratt/warmup_experiment/krun/krun/vm_defs.py", line 629, in _check_truffle_enabled
    "jruby_check_graal_enabled.rb", force_dir=VM_SANITY_CHECKS_DIR)
  File "/home/ltratt/warmup_experiment/krun/krun/util.py", line 250, in spawn_sanity_check
    _ = check_and_parse_execution_results(stdout, stderr, rc)
  File "/home/ltratt/warmup_experiment/krun/krun/util.py", line 216, in check_and_parse_execution_results
    key_set = set(json_data.keys())
AttributeError: 'list' object has no attribute 'keys'
Makefile:58: recipe for target 'bench-startup-no-reboots' failed
make: *** [bench-startup-no-reboots] Error 1
bencher3%
```

### PR DESCRIPTION
**Pull request by snim2**
Fixes #181

### CODE REVIEW

File: startup_runners/startup_runner.rb [hunk 1]
```diff
@@ -1,3 +1,5 @@
+# NOTE: you need to provide krun_measure() and krun_get_wallclock().
+
 # defined this way so we don't measure the conditional platform check.
 if /linux/ =~ RUBY_PLATFORM then
     def clock_gettime_monotonic()
```
Comment (vext01) [inline startup_runners/startup_runner.rb]: I think this might be dead?

**Comment by vext01** [comment]
Both the inner and outer runners are using `krun_measure(1)`. This means we end up doing something like this:

```
        data->wallclock = krun_clock_gettime_monotonic();                       
        for (int core = 0; core < krun_num_cores; core++) {                     
            data->core_cycles[core] = krun_read_core_cycles(core);              
            data->aperf[core] = krun_read_aperf(core);                          
            data->mperf[core] = krun_read_mperf(core);                          
        }                                                                       
        krun_check_mdata();

        // VM starts up here

        data->wallclock = krun_clock_gettime_monotonic();                       
        for (int core = 0; core < krun_num_cores; core++) {                     
            data->core_cycles[core] = krun_read_core_cycles(core);              
            data->aperf[core] = krun_read_aperf(core);                          
            data->mperf[core] = krun_read_mperf(core);                          
        }                                                                       
        krun_check_mdata();
```

In other words, we end up measuring taking un-needed msr readings on top of startup. We also measure the integrity check.

At the very least, the outer runner should use index 0, but I wonder if it would be better to expose `krun_clock_gettime_monotonic` and use it directly?

File: startup.krun [hunk 2]
```diff
@@ -42,6 +42,7 @@ def patch_runner(vm_def):
         assert "iterations_runner." in vm_def.iterations_runner
         _, suffix = vm_def.iterations_runner.rsplit(".", 1)
         vm_def.iterations_runner = os.path.join(STARTUP_RUNNER_DIR, "startup_runner." + suffix)
+    vm_def.instrument = False
```
Comment (ltratt) [inline startup.krun:45]: I'm not totally sure if this should be in this PR. I'm happy to be convinced though!

**Comment by snim2** [comment]
We are taking timing readings twice. Once in the outer running and once in the inner runner. Presumably we want to turn measurements off in between, so how do we do that?

File: startup_runners/startup_runner.js [hunk 3]
```diff
@@ -1,3 +1,4 @@
-// NOTE: you need to provide clock_gettime_monotonic.
-print(clock_gettime_monotonic());
-print("], [-1.0, -1.0]]");
+// NOTE: you need to provide krun_measure() and krun_get_wallclock()..
```
Comment (ltratt) [inline startup_runners/startup_runner.js]: Let's delete this comment (the old one wasn't very useful either).

File: startup_runners/startup_runner.php [hunk 4]
```diff
@@ -1,10 +1,11 @@
 <?php
 
 /*
- * Note! You will need to provide an implementation of clock_gettime_monotonic()
+ * Note! You will need to provide an implementation of krun_measure() and krun_get_wallclock()
```
Comment (ltratt) [inline startup_runners/startup_runner.php]: Similarly, let's kill this comment.

File: startup_runners/startup_runner.rb [hunk 5]
```diff
@@ -1,3 +1,5 @@
+# NOTE: you need to provide krun_measure() and krun_get_wallclock().
```
Comment (ltratt) [inline startup_runners/startup_runner.rb]: Similarly, let's kill this comment.

**Comment by ltratt** [comment]
Best to ask @cfbolz about the methodology.

File: startup.krun (hunk 2 above)
Comment (snim2) [reply startup.krun:45]: If PyPy instrumentation is on in `warmup.krun` and we don't turn it off in `startup.krun` then I get a parse error of sorts. So, either we need to decide that we keep this in and all instrumentation is off in the startup runner, or we remove this line and fix the startup runners to work with all the instrumentation we have.

Comment (ltratt) [reply startup.krun:45]: We can consider startup and instrumentation modes to be mutually exclusive. Maybe this should be an `assert` or something? [I'm not saying I have the answer here... just trying to work out what the right thing is!]

Comment (ltratt) [inline startup.krun]: We can consider startup and instrumentation modes to be mutually exclusive. Maybe this should be an `assert` or something? [I'm not saying I have the answer here... just trying to work out what the right thing is!]

Comment (snim2) [reply startup.krun:45]: I'm not sure what we want here either. Maybe @vext01 has a view?

**Comment by vext01** [comment]
Like I say, we only care about wallclock time, so maybe we should call `krun_clock_gettime_monotonic()` directly and thus skip the other measurements.

**Comment by snim2** [comment]
OK, I will push a commit which does that...

**Comment by snim2** [comment]
OK, what I have now works for every language except Java, and every VM except HHVM and V8. In all three cases the VMs need to be patched to allow access to `krun_clock_gettime_monotonic()`

**Comment by ltratt** [comment]
If Edd's happy with the current approach (it looks sensible to me, but this isn't my parish), then I'd suggest going ahead on that basis.

**Comment by vext01** [comment]
This looks fine.

I guess do Java next.

Do you need help patching VMs, or shall I do that bit? It can be fiddly.

**Comment by snim2** [comment]
I'm not really sure where to start with patching the VMs, happy to have a go if you give me some step by step instructions...

**Comment by vext01** [comment]
In the interest of getting it done quickly, shall I go ahead and do it?

**Comment by snim2** [comment]
That would be great if you can!

**Comment by vext01** [comment]
sure, no worries. It will take a while, as building hhvm takes an age and a day.

**Comment by ltratt** [comment]
If you're using a vaguely recent master, HHVM now builds with `num_cores`, which speeds it up a _lot_.

**Comment by vext01** [comment]
@ltratt ah yeah.

@snim2 before I go ahead, did you look at the output of a startup run to check it looks saneish?

**Comment by snim2** [comment]
Before we moved to `krun_clock_gettime_monotonic()` I checked that we could produce charts form the output file:

startup_chart

The chart itself isn't very useful, but clearly the output file works with our other scripts.

Should we raise a new issue to write a separate script to make nice plots from the startup results?

**Comment by ltratt** [comment]
We don't, I think, want plots from the startup scripts: we want simple tables with means and confidence intervals. Our working assumption is that they're normally distributed.

**Comment by vext01** [comment]
What is the y-axis on that graph? The clock is measured in seconds. Those values look like garbage, no?

Re plots, what @ltratt said.

**Comment by vext01** [comment]
Probably because krun_get_wallclock returns an interval based on two internal measurements, but we only take one measurement in each inner/outer runner.

This should go away after we use krun_clock_gettime_monotonic.

**Comment by snim2** [comment]
Hmm, this still seems a little odd:

``` python
>>> from warmup.krun_results import read_krun_results_file
>>> data = read_krun_results_file("startup_results.json.bz2")
>>> data["wallclock_times"]["binarytrees:JRubyTruffle:default-ruby"]
[[4559160.027431, 4559163.433388]]
>>> 
```

What units does `krun_clock_gettime_monotonic()` return? Seconds?

**Comment by vext01** [comment]
Yes, but you were not using krun_clock_gettime_monotonic() directly above, right?

**Comment by vext01** [comment]
And that's seconds since boot. We should be computing an interval at some point.

**Comment by snim2** [comment]
The chart was old data, the Python session was new data with `clock_gettime_monotonic()`. Where should an interval be computed?

**Comment by vext01** [comment]
I'm not sure. Let me finish patching the VMs and I'll have a look for you.

**Comment by vext01** [comment]
> OK, what I have now works for every language except Java, and every VM except HHVM and V8. 

Did you mean JRuby btw? We shouldn't need to patch the JVM. Just update the JNI bindings.

**Comment by snim2** [comment]
No, I meant that `clock_gettime_monotonic` needs to be added to the JNI bindings.

**Comment by vext01** [comment]
Oh that's right, JRuby already provides access to the clock.

**Comment by vext01** [comment]
The JNI bindings are easyish btw.

File: startup_runners/outer_startup_runner.c [hunk 6]
```diff
@@ -24,9 +24,11 @@ main(int argc, char **argv)
     argc = argc;
     argv = argv;
 
-    krun_measure(1);
-    start_time = krun_get_wallclock(1);
-    fprintf(stdout, "[[%f,\n", start_time);
+    start_time = krun_clock_gettime_monotonic();
+    fprintf(stdout, "{ \"core_cycle_counts\": [[0, 0], [0, 0], [0, 0], [0, 0]],");
+    fprintf(stdout, "  \"mperf_counts\" : [[0, 0], [0, 0], [0, 0], [0, 0]],");
+    fprintf(stdout, " \"aperf_counts\" : [[0, 0], [0, 0], [0, 0], [0, 0]],");
```
Comment (snim2) [inline startup_runners/outer_startup_runner.c]: Is it worth refactoring this so that we don't assume the CPU has four cores?

Comment (vext01) [reply startup_runners/outer_startup_runner.c]: You can use krun_get_num_cores(), which is exposed in all VMs.

**Comment by vext01** [comment]
VMs patched.

hhvm:

```
vext01@bencher5:~/research/warmup_experiment$ ./work/hhvm/hphp/hhvm/hhvm -a     
Welcome to HipHop Debugger!                                                     
Type "help" or "?" for a complete list of commands.                             

Note: no server specified, debugging local scripts only.                        
If you want to connect to a server, launch with "-h" or use:                    
  [m]achine [c]onnect <servername>                                              

hphpd> echo krun_clock_gettime_monotonic();                                     
4502111.3383156
```

v8:                                                                                

```
vext01@bencher5:~/research/warmup_experiment$ ./work/v8/out/native/d8                                          
V8 version 5.1.281.65                                                           

d8> krun_clock_gettime_monotonic;                                               
function krun_clock_gettime_monotonic() { [native code] }                       
d8> print(krun_clock_gettime_monotonic());                                      
4502287.607179753                                                               
```

**Comment by snim2** [comment]
Thanks, will fix the fake data in the outer loop and test.

**Comment by cfbolz** [comment]
It makes sense to me to try to use the timer only directly, otherwise you always measure the time it takes to read the MSR and other stuff.

About the output file: It is not meant to be directly compatible with krun output files. After all, it just produces a single measurement (not 2000) per process iteration. The reason why you are seeing two numbers like this: `[4559160.027431, 4559163.433388]` is because it simply prints the "just before" and "just inside" times of the process start. So what you have to do is take the difference to get the time it starts up.

**Comment by snim2** [comment]
This is just waiting on me to test it out on bencher2, then I'm expecting we can merge later today

**Comment by vext01** [comment]
Thanks for clarifying @cfbolz

**Comment by snim2** [comment]
@ltratt I have done a full run on bencher2 and everything seems fine, so this is ready for review.

File: startup_runners/outer_startup_runner.c [hunk 7]
```diff
@@ -25,7 +25,9 @@ main(int argc, char **argv)
     argv = argv;
 
     start_time = krun_clock_gettime_monotonic();
-    fprintf(stdout, "{ \"core_cycle_counts\": [], \"mperf_counts\" : [], \"aperf_counts\" : [],");
+    fprintf(stdout, "{ \"core_cycle_counts\": [[0, 0], [0, 0], [0, 0], [0, 0]],");
+    fprintf(stdout, "  \"mperf_counts\" : [[0, 0], [0, 0], [0, 0], [0, 0]],");
```
Comment (ltratt) [inline startup_runners/outer_startup_runner.c]: Two spaces before the `\=`?

Comment (snim2) [reply startup_runners/outer_startup_runner.c]: I think to make sure there is a space after the commas in the output.

File: startup_runners/outer_startup_runner.c [hunk 8]
```diff
@@ -6,28 +6,45 @@
...
 
 #define BENCH_FUNC_NAME "run_iter"
+#define LEN_DATUM 8
 
 int
 main(int argc, char **argv)
 {
     double start_time = -1;
     int result;
+    char fake_datum[LEN_DATUM] = "[0, 0], ";
```
Comment (ltratt) [inline startup_runners/outer_startup_runner.c]: I think there's a buffer overrun here: `LEN_DATUM` is length 8, but the RHS string is of length 9 (with the null terminator). I think this probably impacts other parts of the code too.

File: startup_runners/outer_startup_runner.c [hunk 9]
```diff
@@ -6,28 +6,45 @@
...
     /* silence gcc */
     argc = argc;
     argv = argv;
 
     start_time = krun_clock_gettime_monotonic();
-    fprintf(stdout, "{ \"core_cycle_counts\": [[0, 0], [0, 0], [0, 0], [0, 0]],");
-    fprintf(stdout, "  \"mperf_counts\" : [[0, 0], [0, 0], [0, 0], [0, 0]],");
-    fprintf(stdout, " \"aperf_counts\" : [[0, 0], [0, 0], [0, 0], [0, 0]],");
+    fprintf(stdout, "{ \"core_cycle_counts\": [ %s ],", fake_data);
+    fprintf(stdout, "  \"mperf_counts\" : [ %s ],", fake_data);
```
Comment (ltratt) [inline startup_runners/outer_startup_runner.c]: Double space before `\"`.

**Comment by ltratt** [comment]
OK, various minor comments.

**Comment by snim2** [comment]
Ready to squash?

**Comment by ltratt** [comment]
Let's rebase against master, then I can review. Don't squash though please.

**Comment by ltratt** [comment]
Squash as you see fit, then I can merge.

**Review by ltratt — APPROVED** [review]

**Comment by snim2** [comment]
Splat.

**Pull request merged**